definitions/npm/styled-components_v4.x.x/flow_v0.104.x-/styled-components_v4.x.x.js

```diff
@@ -285,133 +302,21 @@ declare module 'styled-components' {
-    getImportRuleTag() : Tag<any>; // eslint-disable-line flowtype/no-weak-types
-    getTagForId(id : string): Tag<any>; // eslint-disable-line flowtype/no-weak-types
-    hasId(id: string) : boolean;
-    hasNameForId(id: string, name: string) : boolean;
-    deferredInject(id : string, cssRules : string[]) : void;
-    inject(id: string, cssRules : string[], name? : string) : void;
-    remove(id : string) : void;
-    toHtml() : string;
-    toReactElements() : React$ElementType[];
-  }
-
-  declare export class KeyFrames {
-    id : string;
-    name : string;
-    rules : string[];
-
-    constructor(name : string, rules : string[]) : this;
-    inject(StyleSheet) : void;
-    toString() : string;
-    getName() : string;
-  }
-
-  // I think any is appropriate here?
-  // eslint-disable-next-line flowtype/no-weak-types
-  declare export var css : CSSConstructor;
-  declare export var keyframes : KeyFramesConstructor;
-  declare export var createGlobalStyle : CreateGlobalStyleConstructor
-  declare export var ThemeProvider: React$ComponentType<{
-    children?: ?React$Node,
-    theme: mixed | (mixed) => mixed,
-    ...
-  }>
-
-  declare type ThemeProps<T> = {|
-    theme: T
-  |}
-
-  declare type PropsWithTheme<Props, T> = {|
-    ...ThemeProps<T>,
-    ...$Exact<Props>
-  |}
-
-  declare export function withTheme<Theme, Config: {...}, Instance>(Component: React$AbstractComponent<Config, Instance>): React$AbstractComponent<$Diff<Config, ThemeProps<Theme | void>>, Instance>
-
-  declare export type StyledComponent<Props, Theme, Instance> = React$AbstractComponent<Props, Instance> & Class<InterpolatableComponent<Props>>
-
-  declare type StyledFactory<StyleProps, Theme, Instance> = {|
-    [[call]]: TaggedTemplateLiteral<PropsWithTheme<StyleProps, Theme>, StyledComponent<StyleProps, Theme, Instance>>;
-    +attrs: <A: {...}>(((StyleProps) => A) | A) => TaggedTemplateLiteral<
-      PropsWithTheme<{|...$Exact<StyleProps>, ...$Exact<A>|}, Theme>,
-      StyledComponent<React$Config<{|...$Exact<StyleProps>, ...$Exact<A>|}, $Exact<A>>, Theme, Instance>
-    >;
-  |}
-
-  declare type StyledShorthandFactory<V> = {|
-    [[call]]: <StyleProps, Theme>(string[], ...Interpolation<PropsWithTheme<StyleProps, Theme>>[]) => StyledComponent<StyleProps, Theme, V>;
-    +attrs: <A: {...}, StyleProps = {||}, Theme = {||}>(((StyleProps) => A) | A) => TaggedTemplateLiteral<
-      PropsWithTheme<{|...$Exact<StyleProps>, ...$Exact<A>|}, Theme>,
-      StyledComponent<React$Config<{|...$Exact<StyleProps>, ...$Exact<A>|}, $Exact<A>>, Theme, V>
-    >;
-  |}
+  import type {
+    CSSRules,
+    CSSConstructor,
+    KeyFramesConstructor,
+    CreateGlobalStyleConstructor,
+    StyledComponent,
+    Interpolation,
+
+    // "private" types
+    TaggedTemplateLiteral,
+    StyledFactory,
+    StyledShorthandFactory,
+    ThemeProps,
+    PropsWithTheme,
+  } from 'styled-components';
```

Comment:
Did you double-check that importing within the module here does not lose typing? Types imported across libraries are implicitly converted to `any`, and while I don't think that's the case within a libdef, it'd be good to verify.

Reply:
Argh! The tests were passing, so I assumed it works! Dang it. :sob:

You're right, it doesn't work the way I thought it does.

I am a bit confused though, because Flow did tell me I can only use `import type {…} from '…'` syntax in libdefs, to import from other libdefs.


I guess we could move the types to the top-level and have them as `npm$StyledComponents$…`, but I was never a fan of that approach :nauseated_face:; nevermind for now then.

Reply:
Scratch that. It _does_ work. I had a typo in the import; I was too occupied checking the type that I didn't notice it complained about the error during import. Lol. :woman_facepalming:

Reply:
I'll mark this as resolved then

Reply:
Sounds good.